New POW calculation module

Kleshni opened the pull request:
For #1227. It's currently not connected to the PyBitmessage code, I'm going to get it done sooner or later.

It provides the `WorkProver` thread class which manages tasks and schedules them to solvers. There are 4 possible solvers:

- Single-threaded `DumbSolver` in Python. I optimised it so it now works ~2 times faster. In particular, I replaced `hashlib.sha512` with `SHA512` from OpenSSL. Since you need OpenSSL to assemble a message anyway, I think it's safe to use it in a fallback solver.

- Multiprocess based `ForkingSolver` in Python. It works faster too, because it relies on `DumbSolver`.

- Multithreaded `FastSolver` in C. I tried to utilize SIMD, but it gave very little speed up of 7 %, so it still calls OpenSSL.

- And `GPUSolver` in OpenCL.

The library was tested on a 4-core AMD64 Linux with integrated GPU and 45-CPU virtual machines with hyperthreading enabled:
- X86 Linux;
- AMD64 FreeBSD;
- AMD64 OpenBSD with the `bsd.mp` kernel;
- AMD64 NetBSD;
- AMD64 Windows 7.

I failed to install Hackintosh on QEMU, so somebody else has to test it there.

PeterSurda commented:
Have you tried frozen mode on Windows?

Kleshni commented:
No, but `ForkingSolver` is disabled in this case complying to the current behaviour. All file paths (to the C library and OpenCL kernel) are relative to the code directory, whose path must be provided by the calling code.

omkar1117 commented on src/workprover/forkingsolver.py:
```diff
@@ -0,0 +1,106 @@
+import os
```
pep8 validation for the file

omkar1117 commented on src/workprover/forkingsolver.py:
```diff
@@ -0,0 +1,106 @@
+def setIdle():
+    try:
+        import psutil
+
```
remove un necessary spaces

Kleshni commented on src/workprover/Readme.md:
```diff
@@ -0,0 +1,42 @@
+Please keep this module independent from the outside code, so that it can be reused in other applications.
```
It's a Python package with the `__init__.py` file, and it's intended to be imported like `import workprover`. It could be moved to a separate repository, but I think, it's easier and safer to keep it in the same file tree.

Kleshni commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,238 @@
+        if self.statusUpdated is None:
+            return
+
+        if self.solver is None:
```
It's less readable.

Kleshni commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,238 @@
+        self.currentTaskID = None
+
+    def notifyStatus(self):
+        if self.statusUpdated is None:
```
It's less acurate.

`self.statusUpdated` must be either a function or a `None`. If it's 0 or an empty string, it would be a programming error and further call would rise an exception making the error noticable.

Kleshni commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,238 @@
+        self.statusUpdated((self.solverName, parallelism, self.speed))
+
+    def setSolver(self, name, parallelism):
+        if name is None and self.solverName is None:
```
It calls the `self.statusUpdated` callback and the outside code should log this and display in the GUI.

g1itch commented:
Some tests [are failing](https://travis-ci.org/g1itch/PyBitmessage/builds/395814851). It seems to be expected for `TestGPUSolver`, what about the rest?

g1itch commented:
[Strange thing](https://travis-ci.org/g1itch/PyBitmessage/builds/395846736):

> OSError: /home/travis/build/g1itch/PyBitmessage/src/workprover/fastsolver/libfastsolver.so: undefined symbol: SHA512_Update

It seems to be specific to ubuntu:trusty or Travis-CI.

Nevertheless could you please add `pybitmessage.workprover` to `packages` and decorate `TestGPUSolver` like I did?

Kleshni commented:
I have registered in Travis-CI and debugged the issue. It was a very stupid mistake, shared libraries like `-lpthread` and `-lcrypto` should be listed after object files in the linker command line.

Travis-CI also gives an OS X environment, so I tested the module there.

PeterSurda commented:
`packages/pyinstaller/bitmessagemain.spec` needs updating.

g1itch commented on src/workprover/gpusolver.py:
```diff
@@ -0,0 +1,106 @@
+
+        try:
+            import numpy
+            import pyopencl
```
Could you please comment #1283 and maybe change the import order too if you agree?

g1itch commented:
Could you please also add the extension and a workaround for dh_python in Debian: a9955e5?

Kleshni commented:
Maybe just invoke the makefile from `setup.py`?

g1itch commented:
The extension approach is more pythonic and there is no need for makefile. You can see it even in the [tests log](https://travis-ci.org/g1itch/PyBitmessage/builds/397318626).

Kleshni commented:
But there rises up a need for a workaround, while the makefile is needed anyway. And a pythonic extention is not a `ctypes` library, it must use Python C API to be directly importable:
>A C extension for CPython is a shared library (e.g. a .so file on Linux, .pyd on Windows), which exports an *initialization function*.

omkar1117 commented on src/api.py:
```diff
@@ -10,8 +10,11 @@
+import socket
+import threading
 import time
 from binascii import hexlify, unhexlify
```
Please place all the imports in a order

omkar1117 commented on src/api.py:
```diff
@@ -1297,3 +1301,49 @@ def _dispatch(self, method, params):
+            s.shutdown(socket.SHUT_RDWR)
+            s.close()
+        except:
+            pass
```
better to place log instead of exception pass.
I hope it will be dangerous if we have pass statement.

omkar1117 commented on src/bitmessagemain.py:
```diff
@@ -499,10 +381,87 @@ def main():
 
+# See "workprover/Readme.md"
+
+import pyelliptic.openssl
```
Place this imports in the starting of the script.

g1itch commented:
Are you going to connect new POW code soon? This PR still not includes the actual use of this code for doing work.

Kleshni commented:
I almost finished rewriting `class_singleWorker`, but I have a problem with the function for requesting pubkeys. It uses the `retrynumber` column in the database for detemining the TTL and rising it exponentially for every next attempt.

But proper implementation of this scheme gets very complicated if two or more messages need the same pubkey and have different retry numbers. It would be wise to not send two simultaneous getpubkey requests in this case, and to not send it at all if it's already present in the inventory. So there emerges a question on how to handle the `retrynumber` column in such situations, and what to do with it when some other user sends a getpubkey request for a key we also need.

I think we can get rid of this column, I mean use it only for resending the message itself but ignore it for getpubkeys. This would simplify the algorithm: just check there is no requests in the inventory, and send it with a fixed TTL.

The comment near the [line](https://github.com/Bitmessage/PyBitmessage/blob/529559d06a3cd1c89838482acaf0fd16a018a057/src/class_singleWorker.py#L1324) calculating the TTL says, that the 2,5-day period was chosen "fairly arbitrarily". Maybe redefine it to the maximum 28 days? Getpubkey requests are negligibly small and are sent with the default network difficulty, they don't need too much work.

There are currently 13661 getpubkeys in the inventory, but only 84 of them have different tags! The others are duplicates, which I propose to avoid by simplifying the sending algorithm. And there are 40939 different pubkeys, so I conclude that the need for getpubkeys is very small, and there is no big reason to save on short TTLs.

Less frequent resending and not repeating already existing requests can also affect privacy in the positive way.

(Another question is that the current requesting function specifies the object type as 0 in one place and 1 in another, but I think it's certainly a bug.)

omkar1117 commented:
@Kleshni can I have the functional flow also along with the PR, it will be easy to test the internal things please.

Kleshni commented:
Sorry, I don't know what the "functional flow" is. I tried to google, but it seems, that I have to learn much new material to satisfy your request.

>with some snapshots
Snapshots of what?

Maybe I should add doc strings to functions or something like that?

Kleshni commented:
I don't completely understand why the TTL randomization is needed:

```python
TTL += random.randrange(-300, 300)
expiryTime = int(time.time() + TTL)
```

Possibly, it's supposed to hide the origin node of the message because it makes neighbour nodes unable to tell whether the message was generated right now or its retranslation took some time.

But it seems very weak:

1. Maybe it could hide singular messages, but if a node generates messages continuously (like the time service broadcast), the expected value of the random addition approaches -0,5 - a constant.

2. An attacker can guess the original TTL, which is usually a multiple of 3600 seconds. If the expiry time of a received object is `currentTime + originalTTL + 299`, then it's obvious, that the random addition was 299 seconds, it can't be more. In this case it hides nothing.

3. TTL affects POW difficulty, POW difficulty affects the nonce value, the nonce value is known to everybody. It's another way to leak the real sending time.

I don't know how to solve the first two problems, but the the last can be solved by randomizing the expiry time instead:

```
TTL += 599
expiryTime = int(time.time() + TTL - random.randrange(600))
```

So I propose to change the code this way.

Kleshni commented:
I deleted the file `bitmessageqt/settings.py`. It was originally generated by the `pyuic4` utility from `bitmessageqt/settings.ui` and was not actually needed, because `*.ui` files can be used directly without conversion to `*.py`.

The generated files must remain untouched because they need a regeneration on every change to the source files, but `settings.py` was manually modified one day. This prevented easy modification of the settings window, so I solved this by moving the changes out from this file and deleting it.

Kleshni commented:
The same for the main window.

omkar1117 commented on src/bitmessageqt/__init__.py:
```diff
@@ -584,9 +582,51 @@ def rerenderTabTree(self, tab):
+        self.ui.horizontalSplitter_6.handle(1).setEnabled(False)
+        self.ui.horizontalSplitterSearch.handle(1).setEnabled(False)
+        self.ui.verticalSplitter.handle(1).setEnabled(False)
+        self.ui.verticalSplitter_2.handle(1).setEnabled(False)
```
can't we make it dynamic or iterable?

omkar1117 commented on src/api.py:
```diff
@@ -1297,3 +1301,49 @@ def _dispatch(self, method, params):
+
+# This thread, of which there is only one, runs the API.
+class singleAPI(threading.Thread, helper_threading.StoppableThread):
+    def __init__(self):
```
please add doc string to the class.

omkar1117 commented on src/api.py:
```diff
@@ -1297,3 +1301,49 @@ def _dispatch(self, method, params):
+        self.initStop()
+
+    def stopThread(self):
+        super(singleAPI, self).stopThread()
```
please add doc string to the class

omkar1117 commented on src/bitmessagemain.py:
```diff
@@ -502,10 +384,85 @@ def main():
 
+# See "workprover/Readme.md"
+
+import os
```
please write all the imports on the starting of the file.

omkar1117 commented on src/bitmessagemain.py:
```diff
@@ -502,10 +384,85 @@ def main():
 
 if __name__ == "__main__":
-    main()
+    import multiprocessing
```
Please place all the imports in the starting the file...

Kleshni replied:
These imports must be conditional.

omkar1117 commented on src/class_singleCleaner.py:
```diff
@@ -35,6 +35,45 @@
 
+def resendStaleMessages():
+    staleMessages = sqlQuery("""
+        SELECT "toaddress", "ackdata", "status" FROM "sent"
```
Pep8 validation here.

omkar1117 commented on src/class_sqlThread.py:
```diff
@@ -148,16 +150,55 @@ def run(self):
+
+        # Update status values inherited from old versions
+
+        self.cur.execute("""UPDATE "sent" SET "status" = 'msgsent' WHERE "status" == 'sentmessage';""")
```
pep8 validation here.

omkar1117 commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,253 @@
+import Queue
```
please place all the imports in a alphabetic order.

omkar1117 commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,253 @@
+        self.statusUpdated(Status(self.solverName, status, self.speed, len(self.tasks), self.totalDifficulty))
+
+    def setSolver(self, name, configuration):
+        if name is None and self.solverName is None:
```
if not name and not self.solverName:
    pass

omkar1117 commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,253 @@
+        elif name == self.solverName:
+            self.solver.setConfiguration(configuration)
+        else:
+            if self.solver is not None:
```
if self.solver:
   ….

omkar1117 commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,253 @@
+                self.solverName = None
+                self.solver = None
+
+            if name is not None:
```
if name:
 ….

Kleshni commented:
I deleted the `getMessageDataByDestination{Hash,Tag}` methods because they were undocumented and didn't work at all, so I concluded that no-one ever used them. They are now replaced by `getRawObject` and `listRawObjects`.

The same for `disseminatePreEncryptedMsg` and `disseminatePubkey`.

g1itch commented on src/workprover/__init__.py:
```diff
@@ -0,0 +1,262 @@
+                    self.solverName = name
+                    self.solver = self.availableSolvers[name]
+                    self.solver.setConfiguration(configuration)
+        except GPUSolverError:
```
undefined name 'GPUSolverError'

g1itch commented:
I got 3 additional threads (`workprover.WorkProver` instances) which don't stop when I run `src/pybitmessage.py -t`:
https://travis-ci.org/g1itch/PyBitmessage/builds/411231409

This patch helped:
```patch
diff --git a/src/singleworker.py b/src/singleworker.py
index 9fa1391c..e1eeda3f 100644
--- a/src/singleworker.py
+++ b/src/singleworker.py
@@ -270,17 +270,19 @@ def setBestSolver():
 
 setBestSolver()
 
+
 class singleWorker(threading.Thread, helper_threading.StoppableThread):
     name = "singleWorker"
 
     def __init__(self):
-        super(self.__class__, self).__init__()
+        super(self.__class__, self).__init__(name="singleWorker")
 
         self.initStop()
 
     def stopThread(self):
         queues.workerQueue.put(("stopThread", "data"))
 
+        workProver.commandsQueue.put(("shutdown", ))
         super(self.__class__, self).stopThread()
 
     def run(self):
```

Kleshni commented:
Any plans on merging?

PeterSurda commented:
@Kleshni I haven't looked at it yet, but it first needs to pass all the code quality checks and all commits need to be signed (you can squash commits if it helps).

However, others gave good feedback so chances are good.

PeterSurda commented:
Also, please split this into multiple patches. The PoW, the API and the other changes should be separate. You'll have better chances of getting it merged that way too.

PeterSurda commented:
@Kleshni would it be ok if I assigned someone else to the task to help you with cleaning it up? You'd have to give them write access to your repo.

Kleshni commented:
OK.

PeterSurda commented:
@Kleshni please give @omkar1117 write access to the POW branch. this way your commits will be preserved and you can still get paid through tip4commit.

Kleshni commented:
I have sent him an invite yesterday but forgot to notify you :eyes:

omkar1117 commented:
@Kleshni could you please resolve the conflicsts please.

PeterSurda commented:
@Kleshni Omkar says he doesn't have write access. He could fork it into his own repo, but since this PR is already open I would prefer to continue this way. Can you check and let me know?

Kleshni commented:
This is what I see in the repository settings:
![Screenshot](https://files.catbox.moe/i8o2jh.png)
